packages/charts/react-charting/src/components/DeclarativeChart/PlotlySchemaAdapter.ts

```diff
@@ -566,23 +575,94 @@ export const transformPlotlyJsonToHeatmapProps = (input: PlotlySchema): IHeatMap
   let zMin = Number.POSITIVE_INFINITY;
   let zMax = Number.NEGATIVE_INFINITY;
 
-  (firstData.x as Datum[])?.forEach((xVal, xIdx: number) => {
-    firstData.y?.forEach((yVal: any, yIdx: number) => {
-      const zVal = (firstData.z as number[][])?.[yIdx]?.[xIdx];
+  // TODO: Add support for date axes
+  if (firstData.type === 'histogram2d') {
+    const isStringArray = (arr: any) => isArrayOfType(arr, (value: any) => typeof value === 'string');
+    const findBinIndex = (
+      bins: string[][] | Bin<number, number>[],
+      value: string | number | null | undefined,
+      isString: boolean,
+    ) => {
+      if (typeof value === 'undefined' || value === null) {
+        return -1;
+      }
+
+      return isString
+        ? (bins as string[][]).findIndex(bin => bin.includes(value as string))
+        : (bins as Bin<number, number>[]).findIndex(bin => (value as number) >= bin.x0! && (value as number) < bin.x1!);
+    };
+    const getBinSize = (bin: Bin<number, number>) => bin.x1! - bin.x0!;
+    const getBinCenter = (bin: Bin<number, number>) => (bin.x1! + bin.x0!) / 2;
+
+    const isXString = isStringArray(firstData.x);
+    const isYString = isStringArray(firstData.y);
+    const xBins = createBins(firstData.x, firstData.xbins?.start, firstData.xbins?.end, firstData.xbins?.size);
+    const yBins = createBins(firstData.y, firstData.ybins?.start, firstData.ybins?.end, firstData.ybins?.size);
+    const zBins: number[][][] = yBins.map(() => xBins.map(() => []));
+    let total = 0;
+
+    firstData.x?.forEach((value, index) => {
+      const xVal = value as string | number | null;
+      const yVal = firstData.y?.[index] as string | number | null | undefined;
+      const xBinIdx = findBinIndex(xBins, xVal, isXString);
+      const yBinIdx = findBinIndex(yBins, yVal, isYString);
+
+      if (xBinIdx !== -1 && yBinIdx !== -1) {
+        zBins[yBinIdx][xBinIdx].push((firstData.z?.[index] as number | null | undefined) ?? 1);
+      }
+    });
 
-      heatmapDataPoints.push({
-        x: input.layout?.xaxis?.type === 'date' ? (xVal as Date) : xVal ?? 0,
-        y: input.layout?.yaxis?.type === 'date' ? (yVal as Date) : yVal,
-        value: zVal,
-        rectText: zVal,
+    const z = zBins.map(row => {
+      return row.map(bin => {
+        const zVal = calculateHistFunc(firstData.histfunc, bin);
+        total += zVal;
+        return zVal;
       });
+    });
 
-      if (typeof zVal === 'number') {
-        zMin = Math.min(zMin, zVal);
-        zMax = Math.max(zMax, zVal);
-      }
+    xBins.forEach((xBin, xIdx) => {
+      yBins.forEach((yBin, yIdx) => {
```

Comment:
yBins value might not be available for react-charting till a new version of chart-utilities is published.

Reply:
The `ybins` property might be present in the data, so I added its type in the Plotly schema (based on the [Histogram2d trace reference](https://plotly.com/javascript/reference/histogram2d/#histogram2d-ybins)) to avoid TypeScript errors.